cou/apps/base.py

```diff
@@ -92,31 +61,21 @@ class OpenStackApplication:
     :raises RunUpgradeError: When an upgrade fails.
     """
 
-    # pylint: disable=too-many-instance-attributes
-
-    name: str
-    status: ApplicationStatus
-    config: dict
-    model: COUModel
-    charm: str
-    machines: dict[str, COUMachine]
-    units: list[ApplicationUnit] = field(default_factory=lambda: [])
     packages_to_hold: Optional[list] = field(default=None, init=False)
     wait_timeout: int = field(default=DEFAULT_WAITING_TIMEOUT, init=False)
     wait_for_model: bool = field(default=False, init=False)  # waiting only for application itself
 
     def __post_init__(self) -> None:
         """Initialize the Application dataclass."""
         self._verify_channel()
-        self._populate_units()
 
     def __hash__(self) -> int:
         """Hash magic method for Application.
 
         :return: Unique hash identifier for Application object.
         :rtype: int
         """
-        return hash(f"{self.name}{self.charm}")
+        return hash(f"{self.name}({self.charm})")
 
     def __eq__(self, other: Any) -> bool:
```

Comment:
is it still necessary to have `hash` and `eq` if now the dataclass is frozen?

Reply:
Unfortunately, they are needed, since class is frozen.

```bash
Full execution log: '/home/ubuntu/.local/share/cou/log/cou-20240215220116.log'  
Connected to 'zaza-ce1e95799ba2' ✔
Analyzing cloud... \2024-02-15 22:01:16 [ERROR] Unexpected error occurred.
2024-02-15 22:01:16 [ERROR] unhashable type: 'dict'
```